pytools/cache.py

```diff
@@ -0,0 +1,273 @@
+"""
+This module serves as an interface for memcached.
+
+Module needs to be initialized before use! See init() function.
+
+It basically is as a wrapper for pymemcache package with additional
+functionality and tweaks specific to serve our needs.
+
+There's also support for namespacing, which simplifies management of different
+versions of data saved in the cache. You can invalidate whole namespace using
+invalidate_namespace() function. See its description for more info.
+
+More information about memcached can be found at https://memcached.org/.
+"""
+from pymemcache.client.hash import HashClient
+import hashlib
+import msgpack
+import datetime
+
+_mc = None  # type: HashClient
+_glob_namespace = None  # type: bytes
+
+# FIXME(roman): There's potentially a big problem with storing namespace versions in
+# the same place (memcached). These can get removed if cache becomes too big. It
+# might be a better idea to offload them into a separate system (like redis).
+
+SHA1_LENGTH = 40
+MAX_KEY_LENGTH = 250
+CONTENT_ENCODING = "utf-8"
+
+
+def init(servers, namespace):
+    """Initializes memcached client. Needs to be called before use.
+
+    Args:
+        servers: List of strings with memcached server addresses (host:port).
+        namespace: Global namespace that will be prepended to all keys.
+    """
+    global _mc, _glob_namespace
+    _mc = HashClient(
+        servers=servers,
+        serializer=_serializer,
+        deserializer=_deserializer,
+    )
+    _glob_namespace = namespace + ":"
+    _glob_namespace = _glob_namespace.encode("ascii")
+    if len(_glob_namespace) + SHA1_LENGTH > MAX_KEY_LENGTH:
+        raise ValueError("Namespace is too long.")
+
+
+def set(key, val, time=0, namespace=None):
+    """Set a key to a given value.
+
+    Args:
+        key: Key of the item.
+        val: Item's value.
+        time: The time after which this value should expire, either as a delta
+            number of seconds, or an absolute unix time-since-the-epoch value.
+            If set to 0, value will be stored "forever".
+        namespace: Optional namespace in which key needs to be defined.
+
+    Returns:
+        True if stored successfully.
+    """
+    if _mc is None: return
+    return set_many({key: val}, time, namespace)
+
+
+def get(key, namespace=None):
+    """Retrieve an item.
+
+    Args:
+        key: Key of the item that needs to be retrieved.
+        namespace: Optional namespace in which key was defined.
+
+    Returns:
+        Stored value or None if it's not found.
+    """
+    if _mc is None: return
+    result = get_many([key], namespace)
+    return result.get(key)
+
+
+def delete(key, namespace=None):
+    """Delete an item.
+
+    Args:
+        key: Key of the item that needs to be deleted.
+        namespace: Optional namespace in which key was defined.
+
+    Returns:
+          True if deleted successfully.
+    """
+    if _mc is None: return
+    return delete_many([key], namespace) == 1
```

Comment:
Oh, I guess `True == 1` is legal in Python. But I thought it might be counting the number of deleted keys at first.